### PR TITLE
Add swarm mode support

### PR DESCRIPTION
**Pull request by ketan**
This was based off #491, with a few changes -
* all commits were squashed into one
* the `SwarmModeDockerClient` and `DefaultSwarmModeDockerClient` subclasses
  were removed and pushed into `DockerClient` and `DefaultDockerClient`

### CODE REVIEW

**Comment by ketan** [comment]
I'm at a loss to understand why checkstyle is failing even on master. But the build is apparently green. https://travis-ci.org/spotify/docker-client/branches

Does someone have any pointers?

**Comment by johnflavin** [comment]
I looked at your latest builds. Checkstyle is passing, so I guess you managed to fix whatever it found distasteful.

But all the tests are failing. Like, every one. I think it could be the unix socket connection. All the test failures are trying to connect to `unix://localhost:80`. Typically the socket connections are to `unix:///var/run/docker.sock`. 

(Total speculation follows:) are you testing this code locally with your docker daemon open to http connections? Because if I remember correctly the daemon on the test machines is not; the only way to connect is `/var/run/docker.sock`.

**Comment by ketan** [comment]
Pretty sure I did not change any setup, in fact I'm on a Mac, so nothing
running on port 80. I suspect it is Travis doing something funny.

Will investigate further.

**Comment by ketan** [comment]
Allright, the build is now green, however the coverage is down because of the plethora of classes that represent the JSON data.

If all is good, can I expect a 5.0.3 release?

**Comment by ketan** [comment]
BTW — support for `docker node` and friends is not part of this PR, if you're OK merging this PR, I'll submit another PR for `docker node` support.

File: src/main/java/com/spotify/docker/client/messages/mount/Driver.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (ketan) [inline src/main/java/com/spotify/docker/client/messages/mount/Driver.java]: Looking at the api docs, for a few references to where this class is used, it appears that the class is just a generic thing that describes a driver name, and options for a variety of things, including networks and tasks.

**Comment by ketan** [comment]
@davidxia — I've changed all timestamp fields from `String` to `java.util.Date`.

> In src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java:
> 
> > @@ -3041,6 +3064,198 @@ public void testIpcMode() throws Exception {
>  
>    }
>  
> +  @Test
> +  public void testInspectSwarm() throws Exception {
> +    requireDockerApiVersionAtLeast("1.24", "swarm support");
> +
> +    final Swarm swarm = sut.inspectSwarm();
> +    assertThat(swarm.createdAt(), is(not(isEmptyOrNullString())));
> Should we check not null for the other fields? Seems like we should check for ID at least.
> 
> 

I'm not sure what "value" asserting on fields adds, esp for code that appears to declarative in nature. I've added a test to assert the swarm `id` attribute as you requested.

**Comment by ketan** [comment]
Also, I could not help but notice that the indentation in most of the code is off. Some files are indented using 2 spaces, others with 4 spaces. What's the deal?

**Comment by ketan** [comment]
All tests are green now

**Comment by mattnworb** [comment]
Since a chunk of the code in this PR was originally contributed by @derekadams, is there any way to maintain that info in the commit author field?

**Comment by ketan** [comment]
I'm not sure how to go about putting multiple authors in the author field. I'm happy to attribute the entire commit to him, I don't care so much about my attribution.

Let me know what you'd prefer, I'd really like to get this merged ASAP :)

**Comment by johnflavin** [comment]
This looks like it will do the trick: http://stackoverflow.com/a/31669789/1288474

**Comment by ketan** [comment]
@johnflavin — Done, travis is green.

**Comment by ketan** [comment]
Bump!

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 2]
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
...
+    final WebTarget resource = resource().path("services");
+    return request(GET, SERVICE_LIST, resource, resource.request(APPLICATION_JSON_TYPE));
+  }
+
+  @Override
+  public List<Service> listServices(Service.Criteria criteria)
+          throws DockerException, InterruptedException {
+    assertAPIVersionIsAbove("1.24");
+    WebTarget resource = resource().path("services");
+    final Map<String, List<String>> filters = new HashMap<String, List<String>>();
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: You can use the diamond type `<>` here. The pom.xml specifies Java 1.7.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 3]
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
...
   @Override
+  public Swarm inspectSwarm() throws DockerException, InterruptedException {
+    assertAPIVersionIsAbove("1.24");
+
+    final WebTarget resource = resource().path("swarm");
+    return request(GET, Swarm.class, resource, resource.request(APPLICATION_JSON_TYPE));
+  }
+
+  @Override
+  public ServiceCreateResponse createService(ServiceSpec spec, ServiceCreateOptions options)
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: nitpick: Can you use `final` for all the parameter arguments in this class?

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 4]
```diff
@@ -1539,6 +1552,163 @@ public LogStream execStart(final String execId, final ExecStartParameter... para
...
+        case 404:
+          throw new ServiceNotFoundException(serviceId);
+        default:
+          throw e;
+      }
+    }
+  }
+
+  @Override
+  public void updateService(String serviceId, Long version, ServiceSpec spec)
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Same as comment here https://github.com/spotify/docker-client/pull/499/files#diff-163ee817c343ffd4cbc08d4c3412a8a4R1563

File: src/main/java/com/spotify/docker/client/messages/swarm/Service.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        public String getServiceName() {
+            return serviceName;
+        }
+
+        public void setServiceName(String serviceName) {
+            this.serviceName = serviceName;
+        }
+    }
+
+    public static class CriteriaBuilder {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/swarm/Service.java]: For consistency, can you use the builder pattern in `ContainerConfig`? Also make sure the access modifiers for classes and methods are as strict as they can be.

File: src/main/java/com/spotify/docker/client/messages/swarm/Task.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        public String getDesiredState() {
+            return desiredState;
+        }
+
+        public void setDesiredState(String desiredState) {
+            this.desiredState = desiredState;
+        }
+    }
+
+    public static class CriteriaBuilder {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/swarm/Task.java]: For consistency, can you use the builder pattern in `ContainerConfig`? Also make sure the access modifiers for classes and methods are as strict as they can be.

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 7]
```diff
@@ -3267,6 +3290,199 @@ public void testOomScoreAdj() throws Exception {
...
+
+    final Task task = sut.listTasks().get(1);
+
+    final List<Task> tasksWithId = sut.listTasks(Task.find().withTaskId(task.id()).build());
+
+    assertThat(tasksWithId.size(), is(1));
+    assertThat(tasksWithId.get(0), equalTo(task));
+  }
+
+  private ServiceSpec createServiceSpec(String serviceName) {
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: nitpick: final for parameter.

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 8]
```diff
@@ -3267,6 +3290,199 @@ public void testOomScoreAdj() throws Exception {
...
+  }
+
+  @Test
+  public void testInspectTask() throws Exception {
+    requireDockerApiVersionAtLeast("1.24", "swarm support");
+
+    final ServiceSpec spec = createServiceSpec(randomName());
+    assertThat(sut.listTasks().size(), is(0));
+    final ServiceCreateResponse response = sut.createService(spec, new ServiceCreateOptions());
+    Thread.sleep(2000); // to give it a while to spin containers
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: I'm not a fan of these sleeps. We should use something like https://github.com/awaitility/awaitility, but since we have them littered throughout the tests already, we can do that in another PR.

**Comment by davidxia** [comment]
@ketan Thanks so much for your patience. I took a closer look. I think we should be good to once you address my comments.

File: src/main/java/com/spotify/docker/client/messages/swarm/RestartPolicy.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.util.Objects;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class RestartPolicy {
+
+    public static final String RESTART_POLICY_NONE = "none";
```
Comment (ketan) [inline src/main/java/com/spotify/docker/client/messages/swarm/RestartPolicy.java]: The docs here https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/create-a-service specify just these 3 policies.

**Comment by davidxia** [comment]
replaced by https://github.com/spotify/docker-client/pull/514
